### ISSUE TITLE
Allow *not* to set SC_DISABLE_SPEEDY cleanly

### ISSUE THREAD

**Issue by dimaqq** [styled-components/styled-components#3211]
Since #3166 got rudely closed, let me try again.

#### What is going on

`styled-components` support build-time environment variables `SC_ATTR`, `SC_DISABLE_SPEEDY` and aliases.

There is a bug in create-react-app / react-scripts / webpack where environment variables that are referenced but not defined are not optimised away.

Most users (?) do not set `REACT_APP_SC_*` variables, triggering this bug.

Thus the entire† environment variable block is included.

† `create-react-app` cleans the shell environment, only variables starting with `REACT_APP_` are allowed, and apparently some predefined values.

#### Why is it bad

* the built package is larger
* there's a chance of leaking something†

† For example, if someone builds their project with `env REACT_APP_SECRET_TOKEN=4242 yarn build` that value is now included in the built package.

#### References

`SC_ATTR` was introduced long time ago: https://github.com/oliverlaz/styled-components/blame/2b1d01e889c7e7259c674a2c61c585419a9a28f9/packages/styled-components/src/constants.js

https://github.com/styled-components/styled-components/pull/2501 added the `REACT_APP_` prefix to both variables

https://github.com/styled-components/styled-components/issues/3166 (previous bug report for this issue)

https://github.com/facebook/create-react-app/issues/8961 (upstream bug report)

https://github.com/dimaqq/sc-3166-demo (reproducer)

#### Without `styled-components`

All good, no `NODE_ENV` or `WDS_SOCKET_HOST` in the build project.

#### With `styled-components`

The built project (beautified) contains the following code:

```js
            var b = "undefined" !== typeof e && (Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0
                }).REACT_APP_SC_ATTR || Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0
                }).SC_ATTR) || "data-styled",
                w = "undefined" !== typeof window && "HTMLElement" in window,
                k = "boolean" === typeof SC_DISABLE_SPEEDY && SC_DISABLE_SPEEDY || "undefined" !== typeof e && (Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0
                }).REACT_APP_SC_DISABLE_SPEEDY || Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0
                }).SC_DISABLE_SPEEDY) || !1,
                x = function() {
                    return n.nc
                };
```

#### With `styled-components` with environment variable

Since the upstream bug is triggered when the referenced environment variable is not set, actually setting it (e.g. empty) should work around the bug.

For example, building the project using `env REACT_APP_SC_ATTR="zz" yarn build` reduces the injected environment blocks to:

```js
            var b = "undefined" !== typeof e ? "zz" : "data-styled",
                w = "undefined" !== typeof window && "HTMLElement" in window,
                k = "boolean" === typeof SC_DISABLE_SPEEDY && SC_DISABLE_SPEEDY || "undefined" !== typeof e && (Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0,
                    REACT_APP_SC_ATTR: "zz"
                }).REACT_APP_SC_DISABLE_SPEEDY || Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0,
                    REACT_APP_SC_ATTR: "zz"
                }).SC_DISABLE_SPEEDY) || !1,
                x = function() {
                    return n.nc
                };
```

Setting more environment variables `env REACT_APP_SC_ATTR="zz" REACT_APP_SC_DISABLE_SPEEDY="" yarn build` reduces the injection further:

```js
            var b = "undefined" !== typeof e ? "zz" : "data-styled",
                w = "undefined" !== typeof window && "HTMLElement" in window,
                k = "boolean" === typeof SC_DISABLE_SPEEDY && SC_DISABLE_SPEEDY || "undefined" !== typeof e && Object({
                    NODE_ENV: "production",
                    PUBLIC_URL: "",
                    WDS_SOCKET_HOST: void 0,
                    WDS_SOCKET_PATH: void 0,
                    WDS_SOCKET_PORT: void 0,
                    REACT_APP_SC_DISABLE_SPEEDY: "",
                    REACT_APP_SC_ATTR: "zz"
                }).SC_DISABLE_SPEEDY || !1,
                x = function() {
                    return n.nc
                };
```

Note that the unprefixed `SC_DISABLE_SPEEDY` is filtered out by `react-scripts` and is not passed to webpack.

#### Supply truthy values to environment variables

`env REACT_APP_SC_ATTR="1" REACT_APP_SC_DISABLE_SPEEDY="1" yarn build` eliminates the injected blocks, at the cost of "speedy" / CSSOM API.

#### Environment variable value `false` is not falsy

`env REACT_APP_SC_ATTR=dummy REACT_APP_SC_DISABLE_SPEEDY=false yarn build` doesn't work as `"false"` is interpreted as a string:

```js
            var b = "undefined" !== typeof e ? "dummy" : "data-styled",
                w = "undefined" !== typeof window && "HTMLElement" in window,
                k = "boolean" === typeof SC_DISABLE_SPEEDY && SC_DISABLE_SPEEDY || "undefined" !== typeof e && "false" || !1,
                x = function() {
                    return n.nc
                };
```

Here, `k` becomes `"false"`, a string, and it's used later as `useCSSOMInjection: !k`, that is CSSOM injection is disabled.

**Comment by kitten** [comment]
> There is a bug in create-react-app / react-scripts / webpack where environment variables that are referenced but not defined are not optimised away.

I'm not sure why you're under the impression that your issue was rudely closed, but as you're saying here, it isn't solvable on our side.

So unfortunately I'm not seeing any new question, statement, or issue here. I'll have to close this issue as a duplicate.

**Comment by dimaqq** [comment]
Okay, I guess have to live with it.

Just to be clear, I don't want to disable CSSOM.

The way I see this, pardon for dragging this on:
* this library explicitly supports CRA
* minority use case (CSSOM disabled) is supported cleanly
* majority use case (defaults settings) is... unkempt, the env block is included 4 times in build.

**Comment by MaskedHawk** [comment]
Any news about this issue ? 

it should not have been close since it present a real issue of security and remove the feasibility of code splitting just by adding some style with your library.

in constant.ts by testing the process env to be `no undefined, a boolean, and not false` and cast into Boolean you expose all react env and expose it in the main chunk during bundling.

You should concider to remove this part or find a new way to ensure, that your lib still safe to use.

**Comment by amadebusuyi** [comment]
I don't think this should be closed.

If an environment variable is missing, `undefined` is way better than the compromise of exposing all react env in the main chunk during bundling.

Any chance this will be re-opened and fixed?
